File: infoundr-site-main/src/frontend/src/types/accelerator.ts
/*import type { Principal } from '@dfinity/principal';

export interface Accelerator {
  id: Principal;
  graduated_startups: number;
  active_startups: number;
  logo: [] | [Array<Uint8Array | number[]>];
  name: string;
  email: string;
  website: string;
  team_members: TeamMember[];
  total_startups: number;
  recent_activity: Activity[];
  email_verified: boolean;
  invites_sent: number;

}

export interface TeamMember {
  status: MemberStatus;
  principal: [] | [Principal];
  token: [] | [string];
  role: Role;
  email: string;
}

export interface Activity {
  activity_type: ActivityType;
  description: string;
  timestamp: bigint;
}

export type ActivityType = 
  | { Graduated: null }
  | { UpdatedPitchDeck: null }
  | { Joined: null }
  | { Other: string }
  | { AcceleratorCreated: null }
  | { SentInvite: null }
  | { MissedMilestone: null };

export type MemberStatus = { Active: null } | { Pending: null };

export type Role = { ProgramManager: null } | { Viewer: null } | { SuperAdmin: null } | { Admin: null };

export interface AcceleratorStats {
  total_startups: number;
  invites_sent: number;
  active_startups: number;
  graduated_startups: number;
} */
import type { Principal } from '@dfinity/principal';

export interface Accelerator {
  id: Principal;
  graduated_startups: number;
  active_startups: number;
  logo: [] | [Array<Uint8Array | number[]>];
  name: string;
  email: string;
  website: string;
  team_members: TeamMember[];
  total_startups: number;
  recent_activity: Activity[];
  email_verified: boolean;
  invites_sent: number;
}

export interface TeamMember {
  status: MemberStatus;
  principal: [] | [Principal];
  principalId?: string; // ✅ Flattened Principal string for frontend use
  token: [] | [string];
  role: Role;
  roleString?: RoleUnion; // ✅ Human-readable role string
  email: string;
}

export interface Activity {
  activity_type: ActivityType;
  description: string;
  timestamp: bigint;
}

export type ActivityType =
  | { Graduated: null }
  | { UpdatedPitchDeck: null }
  | { Joined: null }
  | { Other: string }
  | { AcceleratorCreated: null }
  | { SentInvite: null }
  | { MissedMilestone: null };

export type MemberStatus = { Active: null } | { Declined: null } | { Pending: null };

export type Role =
  | { ProgramManager: null }
  | { Viewer: null }
  | { SuperAdmin: null }
  | { Admin: null };

// ✅ For role string mapping (e.g., "Admin", "Viewer")
export type RoleUnion = 'ProgramManager' | 'Viewer' | 'SuperAdmin' | 'Admin';

export interface AcceleratorStats {
  total_startups: number;
  invites_sent: number;
  active_startups: number;
  graduated_startups: number;
}

File: infoundr-site-main/src/backend/src/models/accelerator.rs
use crate::models::stable_principal::StablePrincipal;
use candid::{CandidType, Decode, Encode, Principal};
use ic_stable_structures::{BoundedStorable, Storable};
use serde::{Deserialize, Serialize};
use std::borrow::Cow;

#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum Role {
    SuperAdmin,
    Admin,
    ProgramManager,
    Viewer,
}

#[derive(CandidType, Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
pub enum MemberStatus {
    Active,
    Declined,
    Pending,
}

#[derive(CandidType, Serialize, Deserialize, Clone, Debug)]
pub struct TeamMember {
    pub email: String,
    pub role: Role,
    pub status: MemberStatus,
    pub token: Option<String>,
    pub principal: Option<Principal>,
    pub name: String,
}

#[derive(CandidType, Serialize, Deserialize, Clone, Debug)]
pub struct Accelerator {
    pub id: StablePrincipal,
    pub name: String,
    pub website: String,
    pub email: String,
    pub email_verified: bool,
    pub logo: Option<Vec<Vec<u8>>>,
    pub total_startups: u32,
    pub invites_sent: u32,
    pub active_startups: u32,
    pub graduated_startups: u32,
    pub recent_activity: Vec<Activity>,
    pub team_members: Vec<TeamMember>,
}

#[derive(CandidType, Serialize, Deserialize, Clone, Debug)]
pub struct Activity {
    pub timestamp: u64,
    pub description: String,
    pub activity_type: ActivityType,
}

#[derive(CandidType, Serialize, Deserialize, Clone, Debug)]
pub enum ActivityType {
    AcceleratorCreated,
    Joined,
    UpdatedPitchDeck,
    SentInvite,
    Graduated,
    MissedMilestone,
    Other(String),
}

impl Storable for Accelerator {
    fn to_bytes(&self) -> Cow<[u8]> {
        Cow::Owned(Encode!(self).unwrap())
    }

    fn from_bytes(bytes: Cow<[u8]>) -> Self {
        Decode!(bytes.as_ref(), Self).unwrap()
    }
}

impl BoundedStorable for Accelerator {
    const MAX_SIZE: u32 = 8192; // Adjust as needed for blob data
    const IS_FIXED_SIZE: bool = false;
} 